src/display/api.js

```diff
@@ -1508,28 +1512,28 @@ var WorkerTransport = (function WorkerTransportClosure() {
         loadingTask._capability.resolve(pdfDocument);
       }, this);
 
-      messageHandler.on('NeedPassword',
-                        function transportNeedPassword(exception) {
-        var loadingTask = this.loadingTask;
-        if (loadingTask.onPassword) {
-          return loadingTask.onPassword(updatePassword,
-                                        PasswordResponses.NEED_PASSWORD);
-        }
-        loadingTask._capability.reject(
-          new PasswordException(exception.message, exception.code));
-      }, this);
+      messageHandler.on('Password', function transportPassword(exception) {
+        this._passwordCapability = createPromiseCapability();
 
-      messageHandler.on('IncorrectPassword',
-                        function transportIncorrectPassword(exception) {
-        var loadingTask = this.loadingTask;
         if (loadingTask.onPassword) {
-          return loadingTask.onPassword(updatePassword,
-                                        PasswordResponses.INCORRECT_PASSWORD);
+          var updatePassword = function (password) {
+            this._passwordCapability.resolve({
+              password: password,
+            });
+          }.bind(this);
+          try {
+            loadingTask.onPassword(updatePassword, exception.code);
+            return this._passwordCapability.promise;
+          } catch (ex) { }
         }
-        loadingTask._capability.reject(
-          new PasswordException(exception.message, exception.code));
+        var error = new PasswordException(exception.message, exception.code);
+        loadingTask._capability.reject(error);
+
+        this._passwordCapability.reject(error);
```

Comment:
can you check for `!loadingTask.onPassword` and returned rejected promise (with PasswordException).